src/analyses/constant_propagator.cpp

```diff
@@ -264,6 +282,22 @@ void constant_propagator_domaint::transform(
 #endif
 }
 
+static void
+replace_typecast_of_bool(exprt &lhs, exprt &rhs, const namespacet &ns)
+{
+  // (int)var xx 0 ==> var xx (_Bool)0 or similar (xx is == or !=)
+  // Note this is restricted to bools because in general turning a widening
+  // into a narrowing typecast is not correct.
+  if(
+    lhs.id() == ID_typecast &&
+    (lhs.op0().type().id() == ID_bool || lhs.op0().type().id() == ID_c_bool))
```

Comment:
I'd rewrite this code a bit to do an early `if(lhs.id() != ID_typecast) return;` and then use `to_typecast_expr` instead of using the generic `.op0()`.